### DISCUSSION TITLE
Tabs and Spaces

### DISCUSSION THREAD

**Discussion by icweaver** [fonsp/Pluto.jl#1824] [Pluto development] [answered]
*Wait don't leave!*

I swear this isn't that kind of post. This is more about sharing code snippets with other folks here on the forum.

## The problem
I could be missing the obvious, but when it comes to just copy and pasting a little snippet into a markdown block here on the site, these kinds of shenanigans tend to happen for me at least:

![gh_render](https://user-images.githubusercontent.com/25312320/149641182-cd4a3c13-bc7e-4418-80d3-9b02542cccfc.png)

I saw that there is a [way to handle](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-github-user-account/managing-user-account-settings/managing-your-tab-size-rendering-preference) how GitHub renders the above, but it has never really seemed to work out when I try it.

To avoid this altogether, I've been doing the maybe not so advisable thing of just converting all of the tabs to spaces before sharing them on a GitHub post. Like from this [recent discussion](https://github.com/fonsp/Pluto.jl/discussions/923#discussioncomment-1976179) for example:

```julia
using PlutoUI
using MarkdownLiteral: @mdx

function labeled_boxes(vec)
    PlutoUI.combine() do Child
        @mdx("""
        $([
            @mdx("$(v): $(Child("var$(v)", CheckBox()))")
            for v ∈ vec
        ])
        """)
    end
end

@bind vars_labeled labeled_boxes([1, 2, 3])

vars_labeled
```
(I went with tabs in that post to avoid what happens next, but in the snippet above they should all be spaces)

The problem that I've been hitting is that copying and pasting the snippet above into a Pluto notebook can lead to something like this happening pretty easily:

https://user-images.githubusercontent.com/25312320/149642986-a5e2f38b-18df-4c42-bec4-0a2cadcfbd77.mp4

The source snippet uses spaces for indents while the Pluto editor uses tabs. This caused Pluto to automatically add those two tabs above when starting a newline, that I then went back and manually re-wrote as 8 spaces in the video clip above to demonstrate the different behaviors in the final rendering. As seen in that example, trying to do something as seemingly innocuous as adding a title can lead to the above as the tabs and spaces become mixed and change the final resulting `String`.

## The question
So, tl;dr, I was wondering if it might be time to start resurrecting these `one day` issues (#450, #586) now that folks are (hopefully) starting to use this discussion space more? Uh, pun not intended

**Comment by MichaelHatherly** [comment]
Mixing tabs and spaces in multiline strings can lead to some possibly weird behaviour regarding where leading spaces/tabs end up, which then causes `CommonMark` to get confused, like above, and parse it all as a code block. You can see the issue manifest itself below:

![image](https://user-images.githubusercontent.com/6144086/149658892-030fcd89-0cb3-442f-9767-177b5edf6327.png)

All those look visually identical, but the resulting `String` are not identical.

**Reply by icweaver** [reply]
Hi Michael, right, thanks! I was hoping to convey your point in that little video clip, but definitely could have been more explicit that the resulting `String`s are different. I've updated the caption for it to now read:

> The source snippet uses spaces for indents while the Pluto editor uses tabs. This caused Pluto to automatically add those two tabs above when starting a newline, that I then went back and manually re-wrote as 8 spaces in the video clip above to demonstrate the different behaviors in the final rendering. As seen in that example, trying to do something as seemingly innocuous as adding a title can lead to the above as the tabs and spaces become mixed and change the final resulting String.

**Reply by icweaver** [reply]
So doing some sort of cleaning, or even just highlighting the tab or space characters I think could go a long way in mitigating potential problems when going back and forth between Pluto and GitHub. This is getting well above my pay grade though, so I could be mistaken about how useful this might be!

**Reply by icweaver** [reply]
By the way, I love how your CommonMark.jl package can handle things like this:
![Screenshot from 2022-01-16 11-06-43](https://user-images.githubusercontent.com/25312320/149667833-de72495a-4213-4273-8e13-b1a2acea2fa1.png)

which I think answers [this comment](https://github.com/fonsp/Pluto.jl/issues/450#issuecomment-865011043) in #450?

**Comment by icweaver** [comment]
Pasting a code block from Pluto into the julia REPL seems to convert a tab indent into 8 spaces for me as well (sharing a screenshot instead of a code block again because of the GitHub rendering issue mentioned above):

![Screenshot from 2022-01-21 10-50-04](https://user-images.githubusercontent.com/25312320/150557368-aba51960-4d10-4de9-98e0-5ccf0aceff34.png)

![Screenshot from 2022-01-21 10-52-20](https://user-images.githubusercontent.com/25312320/150557792-9b8d464f-9502-4433-97af-597953c34346.png)

**Reply by fonsp** [reply]
Did it convert it to 8 spaces, or is your terminal *displaying* the tab as an 8-character space?

**Reply by icweaver** [reply]
It converted it to 8 spaces

**Comment by fonsp** [comment]
It's a shame that Julia is not handling this correctly!

**Comment by fonsp** [comment]
Maybe we can normalize the behaviour by secretly replacing `[start of line][n tabs]` by `[start of line][4*n spaces]` before we give cell code to Julia's parser? But we leave the file intact?

**Reply by fonsp** [reply]
No because then running code in Pluto would give different results than running the same code in Julia.

**Reply by icweaver** [reply]
Yea, I know it's been said to death, but I guess all of these issues in the tab vs. spaces flame war are stemming from the fact that we have these invisible things that can do very visible things. So why not just make them visible to let the user know that shenanery is afoot? For example something like this?

![Screenshot from 2022-01-25 13-32-13](https://user-images.githubusercontent.com/25312320/151037606-23dd1b9b-fe2f-47ed-8eb6-46b00a56e87a.png)
 (ofc much prettier, I can't art)

Then they can modify their behavior accordingly (like: "oh snap, some spaces snuck in, let me replace that with a tab") instead of needing to rely on background functionality to do this for them. That way, both sides walk away unhappy, which I've heard is the sign of a good compromise

**Reply by fonsp** [reply]
This would definitely help, but I think we should keep searching for a solution that does not involve the user. In the end, this is a mistake on the side of Julia and Pluto.

**Reply by icweaver** [reply]
haha, I appreciate you and your alter ego wading into this. I just suggested the highlighting thing as a sort of path of least resistance right now, but I agree that having some tooling to deal with the fuster-cluck that is whitespace characters would be the dream

**Reply by icweaver** [reply]
> No because then running code in Pluto would give different results than running the same code in Julia.

I think this is also already happening though, no?

![Screenshot from 2022-01-25 18-03-58](https://user-images.githubusercontent.com/25312320/151074522-ab1399c0-5c25-4932-8945-c7f260d85e7a.png)

```
julia> md"""
           one tab
           four spaces
       """
  one tab
  four spaces

julia> cm"""
           one tab
           four spaces
       """
   │ one tab
   │ four spaces

```

So would it be possible to just convert to spaces in the clipboard only, while keeping the source intact? So sort of the inverse of your suggestion [here](https://github.com/fonsp/Pluto.jl/discussions/1824#discussioncomment-2045719) for going the other direction

**Reply by icweaver** [reply]
idk, maybe this is something that would just be better to start hashing out during one of the weekly calls?

**Comment by fonsp** [comment]
I was thinking about simply switching Pluto to create 4 spaces instead of tabs when you press `Tab` (since this is more common in code snippets on the web), but this will create even more problems when editing existing notebooks! So what about:

We automatically use tabs or 4 spaces, based on what the rest of the code in that cell uses?

**Reply by fonsp** [reply]
This only solves half of the problem: when you add new code to a copy-pasted snippet that uses spaces. But pasting that snippet into a cell that already has tabs will still cause problems.

**Comment by fonsp** [comment]
Like https://github.com/fonsp/Pluto.jl/discussions/1824#discussioncomment-2044853 but on paste:

We can normalize the behaviour by replacing [start of line][4*n spaces] by [start of line][n tabs] in any code that you paste into a cell.

**Comment by LinusSch** [comment]
> > why not just make them visible
> This would definitely help, but I think we should keep searching for a solution that does not involve the user. In the end, this is a mistake on the side of Julia and Pluto.

The way I see it, the mistake has always been that the two are visually indistinguishable in many cases. I'm not sure we can place the blame anywhere else than in some or other distant past. I'm fairly sure it will never be resolved unless the norm becomes that they can be visually distinguished.

Yes, I do use an advanced text editor that makes it so that I almost never have to think about it. It automagically does the right thing... after *configuration* according to the *preferences* of the environment, thus for many usecases I have been involved (but just once per usecase, if at all!). For Julia, in most cases, the low-friction approach is to normalize indentation to four spaces. But note "in most cases". In general, the only way to not involve the user is to do the wrong thing some of the time (which eventually involves the user, perhaps in a worse way).

Yes I do think it is time to resurrect issue #586.

**Comment by icweaver** [comment, marked as answer]
This new settings update helps a bunch! Thanks so much Fons and the Pluto team! https://github.com/JuliaPluto/Pluto.jl/pull/3556
